### PR TITLE
Streamline size estimates

### PR DESCRIPTION
**Pull request by nnethercote**
Makes things nicer and a tiny bit faster.

r? @wesleywiser

### CODE REVIEW

File: compiler/rustc_middle/src/mir/mono.rs [hunk 1]
```diff
@@ -230,14 +230,21 @@ pub struct CodegenUnit<'tcx> {
...
     size_estimate: usize,
     primary: bool,
     /// True if this is CGU is used to hold code coverage information for dead code,
     /// false otherwise.
     is_code_coverage_dead_code_cgu: bool,
 }
 
+/// Auxiliary info about a `MonoItem`.
+#[derive(Copy, Clone, PartialEq, Debug, HashStable)]
+pub struct MonoItemAux {
```
Comment (wesleywiser) [inline compiler/rustc_middle/src/mir/mono.rs:250]: Nit: I think we would usually name this `MonoItemData`. I don't feel strongly about the current name.

File: compiler/rustc_middle/src/mir/mono.rs [hunk 2]
```diff
@@ -59,12 +59,19 @@ impl<'tcx> MonoItem<'tcx> {
     pub fn size_estimate(&self, tcx: TyCtxt<'tcx>) -> usize {
         match *self {
             MonoItem::Fn(instance) => {
-                // Estimate the size of a function based on how many statements
-                // it contains.
-                tcx.instance_def_size_estimate(instance.def)
+                match instance.def {
+                    // "Normal" functions size estimate: the number of
+                    // statements, plus one for the terminator.
```
Comment (wesleywiser) [inline compiler/rustc_middle/src/mir/mono.rs:64]: Oh, seeing this code again reminded me of something that occurred to me while reading your recent blog post: should we have different size estimates for codegen with and without optimizations enabled? 

I could believe this current estimate is relatively accurate for debug builds but many optimization passes in LLVM are super linear with regards to the number of basic blocks. Perhaps it makes sense to use different estimation strategies when we are optimizing code and not optimizing code?

**Review by wesleywiser — APPROVED** [review]

File: compiler/rustc_middle/src/mir/mono.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_middle/src/mir/mono.rs]: I have renamed it.

File: compiler/rustc_middle/src/mir/mono.rs (hunk 2 above)
Comment (nnethercote) [reply compiler/rustc_middle/src/mir/mono.rs:64]: I have tried a variety of alternative estimation functions. #113047 and #113649 are two that I filed PRs for, and I've tried a bunch more locally. I can't say that I ever really saw meaningful differences between release and debug builds. Cleverer estimation functions have gotten me exactly zero wins at the moment, so I'm inclined to not spend more effort there :)

**Comment by the8472** [comment]
There has been a bors issue that lead to the merge commit of this PR getting purged from master.
You'll have to make a new PR to reapply it.

**Pull request merged**